block/azure/chunkwriting.go

```diff
@@ -0,0 +1,273 @@
+	}
+	return nil
+}
+
+// copyFromReader copies a source io.Reader to blob storage using concurrent uploads.
+// TODO(someone): The existing model provides a buffer size and buffer limit as limiting factors.  The buffer size is probably
+// useless other than needing to be above some number, as the network stack is going to hack up the buffer over some size. The
+// max buffers is providing a cap on how much memory we use (by multiplying it times the buffer size) and how many go routines can upload
+// at a time.  I think having a single max memory dial would be more efficient.  We can choose an internal buffer size that works
+// well, 4 MiB or 8 MiB, and autoscale to as many goroutines within the memory limit. This gives a single dial to tweak and we can
+// choose a max value for the memory setting based on internal transfers within Azure (which will give us the maximum throughput model).
+// We can even provide a utility to dial this number in for customer networks to optimize their copies.
+func copyFromReader(ctx context.Context, from io.Reader, to blockWriter, toIDs blockWriter, toSizes blockWriter, o azblob.UploadStreamToBlockBlobOptions) (string, error) {
+	if err := defaults(&o); err != nil {
+		return "", err
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	cp := &copier{
+		ctx:     ctx,
+		cancel:  cancel,
+		reader:  block.NewHashingReader(from, block.HashFunctionMD5),
+		to:      to,
+		toIDs:   toIDs,
+		toSizes: toSizes,
+		id:      newID(),
+		o:       o,
+		errCh:   make(chan error, 1),
+	}
+
+	// Send all our chunks until we get an error.
+	var err error
+	for {
+		if err = cp.sendChunk(); err != nil {
+			break
+		}
+	}
+	// If the error is not EOF, then we have a problem.
+	if err != nil && !errors.Is(err, io.EOF) {
+		return "", err
+	}
+
+	// Close out our upload.
+	if err := cp.close(); err != nil {
+		return "", err
+	}
+
+	// This part was added
+	// Instead of committing in close ( what is originally done in azblob/chunkwriting.go)
+	// we stage The blockIDs and size of this copy operation in the relevant blockBlobURLs ( cp.toIDs, cp.toSizes)
+	// Later on, in complete multipart upload we commit the file blockBlobURLs with etags as the blockIDs
+	// Then by reading the files we get the relevant blockIDs and sizes
+	etag := "\"" + hex.EncodeToString(cp.reader.Md5.Sum(nil)) + "\""
+	base64Etag := base64.StdEncoding.EncodeToString([]byte(etag))
+
+	// write to blockIDs
+	pd := strings.Join(cp.id.issued(), "\n") + "\n"
+	_, err = cp.toIDs.StageBlock(cp.ctx, base64Etag, strings.NewReader(pd), cp.o.AccessConditions.LeaseAccessConditions, nil, cp.o.ClientProvidedKeyOptions)
+	if err != nil {
+		return "", fmt.Errorf("failed staging part data: %w", err)
+	}
+	// write block sizes
+	sd := strconv.Itoa(int(cp.reader.CopiedSize)) + "\n"
+	_, err = cp.toSizes.StageBlock(cp.ctx, base64Etag, strings.NewReader(sd), cp.o.AccessConditions.LeaseAccessConditions, nil, cp.o.ClientProvidedKeyOptions)
```

Comment:
Suggestion:
Use a struct:
```go
type PartInfo struct {
  blockIDs []string `json:"block_ids"`
  size int `json:"size"`
}
```

Then, have just one file for all the blockInfos, and read it with one method and not two.